src/tokens.rs

```diff
@@ -457,32 +497,54 @@ fn update_finalized_transaction(
 
 // Update the finalized bit on any transactions that are now rooted
 // Return the lowest number of confirmations on the unfinalized transactions or None if all are finalized.
+// Also return the unconfirmed transactions with valid blockhashes.
 fn update_finalized_transactions<T: Client>(
     client: &ThinClient<T>,
     db: &mut PickleDb,
-) -> Result<Option<usize>, Error> {
-    let transaction_data = read_transaction_data(db);
-    let unconfirmed_signatures: Vec<_> = transaction_data
+) -> Result<(Option<usize>, Vec<Transaction>), Error> {
+    let transaction_infos = read_transaction_infos(db);
+    let unconfirmed_transactions: Vec<_> = transaction_infos
         .iter()
-        .filter_map(|(signature, info)| {
+        .filter_map(|info| {
             if info.finalized {
                 None
             } else {
-                Some(*signature)
+                Some(&info.transaction)
             }
         })
         .collect();
+    let unconfirmed_signatures = unconfirmed_transactions
+        .iter()
+        .map(|tx| tx.signatures[0])
+        .collect_vec();
     let transaction_statuses = client.get_signature_statuses(&unconfirmed_signatures)?;
+    let recent_blockhashes = client.get_recent_blockhashes()?;
+
     let mut confirmations = None;
-    for (signature, opt_transaction_status) in unconfirmed_signatures
+    let mut mia_transactions = vec![];
+    for (transaction, opt_transaction_status) in unconfirmed_transactions
         .into_iter()
         .zip(transaction_statuses.into_iter())
     {
-        if let Some(confs) = update_finalized_transaction(db, &signature, opt_transaction_status)? {
-            confirmations = Some(cmp::min(confs, confirmations.unwrap_or(usize::MAX)));
+        match update_finalized_transaction(
+            db,
+            &transaction.signatures[0],
+            opt_transaction_status,
+            &transaction.message.recent_blockhash,
+            &recent_blockhashes,
+        ) {
+            Ok(Some(confs)) => {
+                confirmations = Some(cmp::min(confs, confirmations.unwrap_or(usize::MAX)));
```

Comment:
This is a bit lossy. Instead, we could return the Vec of confirmations and use a MultiProgress bar to display progress on all of them.